src/gallery.js

```diff
@@ -0,0 +1,103 @@
+			current += 1
+		}
+	},
+
+	renderPrevious = () => {
+		const image = document.querySelector( '.wp-gallery-popup-image' ).firstChild,
+			caption = document.querySelector( '.wp-gallery-popup-caption' ),
+			previousButton = document.querySelectorAll( '.wp-gallery-popup-button.previous' )[ 0 ],
+			nextButton = document.querySelectorAll( '.wp-gallery-popup-button.next' )[ 0 ],
+			previous = current - 1
+
+		if ( gallery[ previous ] ) {
+			image.src = gallery[ previous ].src
+			caption.innerHTML = gallery[ previous ].caption ? gallery[ previous ].caption : ''
+			previousButton.style.opacity = gallery[ previous - 1 ] ? '1' : '0.5'
+			nextButton.style.opacity = current === gallery.leght - 1 ? '1' : null
+			current -= 1
+		}
+	},
+
+	showFullscreenGallery = ( event, mediaItems ) => {
+		const fullscreenGallery = document.createElement( 'div' ),
+			closeButton = document.createElement( 'div' ),
+			nextButton = document.createElement( 'div' ),
+			previousButton = document.createElement( 'div' ),
+			imageContainer = document.createElement( 'div' ),
+			image = document.createElement( 'img' ),
+			caption = document.createElement( 'p' ),
+			src = event.target.src
+
+		gallery = mediaItems
+		gallery.forEach( ( image, index ) => {
+			if ( image.src === src ) {
+				current = index
+			}
+		} )
+
+		fullscreenGallery.classList.add( 'wp-gallery-popup' )
+
+		image.src = src
+		imageContainer.appendChild( image )
+		imageContainer.classList.add( 'wp-gallery-popup-image' )
+
+		caption.innerHTML = gallery[ current ].caption ? gallery[ current ].caption : ''
+		caption.classList.add( 'wp-gallery-popup-caption' )
+
+		closeButton.classList.add( 'wp-gallery-popup-button', 'close' )
+		closeButton.addEventListener( 'click', hideFullscreenGallery )
+		nextButton.classList.add( 'wp-gallery-popup-button', 'next' )
+		nextButton.style.opacity = current === gallery.length - 1 ? '0.5' : '1'
+		nextButton.addEventListener( 'click', renderNext )
+		previousButton.classList.add( 'wp-gallery-popup-button', 'previous' )
+		previousButton.style.opacity = current === 0 ? '0.5' : '1'
+		previousButton.addEventListener( 'click', renderPrevious )
+
+		Array.prototype.forEach.call(
+			[ closeButton, nextButton, previousButton, imageContainer, caption ],
+			element => {
+				fullscreenGallery.appendChild( element )
+			}
+		)
+
+		document.body.appendChild( fullscreenGallery )
+
+	},
+
+	getGalleryRow = ( mediaItems ) => {
+		const galleryRow = document.createElement( 'div' )
+		galleryRow.classList.add( 'wikipediapreview-gallery-row' )
+
+		mediaItems.forEach( item => {
+			const image = document.createElement( 'img' )
+
+			image.classList.add( 'wikipediapreview-gallery-image' )
+			image.src = item.src
+			image.addEventListener( 'click', ( e ) => { showFullscreenGallery( e, mediaItems ) } )
+			galleryRow.appendChild( image )
+		} )
+
+		return galleryRow
+	}
+
+export { getGalleryRow, showFullscreenGallery }
```

Comment:
it is interesting to see we have a different approach here to build the view.

**In preview and popup**, we use the way of the template (similar to Backbone .template file), therefore preview.js keeps only the html and the event being stored in the event.js

**In gallery**, we use the fundamental way of building the view + event, therefore the gallery is fully controlled in this way, which also brings some benefit but maintenance cost.

---

These two methods can become harder for us to maintain in the future, and also we don't use any framework. Anyway, this doesn't harm as we are in the prototype stage, do keep this in our mind for future reference.